sycl/doc/design/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,1078 @@
+
+namespace {
+  constexpr sycl::specialization_id<int> same_name{2}:
+  /* application code that referenes ::(unnamed)::same_name */
+  namespace {
+    constexpr sycl::specialization_id<int> same_name{3}:
+    /* application code that referenes ::(unnamed)::(unnamed)::same_name */
+  }
+}
+
+/* application code that references "::same_name" */
+```
+
+In that case we can't use `same_name` for specializing
+`get_spec_constant_symbolic_ID`, because it would be ambiguous reference.
+However, we can do the following trick:
+
+```
+// Content of integration footer for the example above
+
+// For unambiguous references we can generate regular specialization
+template<>
+inline const char *get_spec_constant_symbolic_ID<::same_name>() {
+  return "result of __builtin_sycl_unique_id(::same_name) encoded here";
+}
+
+// For ambiguous references we generate 'shim' functions, which allows us to
+// get an address of a variable within a (possible nested) anonymous namespace
+// without spelling it.
+namespace {
+  namespace __sycl_detail {
+    // This helper is need to get addresses of variables defined within
+    // anonymous namespace.
+    // It is generated for each specialization_id within an anonymous namespace
+    // if there is the same specialization_id defined in global namespace
+    static constexpr decltype(spec_name) __spec_id_shim_0() {
+      // address of ::(unnamed)::same_name;
+      return spec_name;
+    }
+  }
+}
+namespace sycl {
+  namespace detail {
+    // By using 'shim' function were are able to unambiguously refer to a
+    // variable within an anonymous namespace
+    template<>
+    inline const char *get_spec_constant_symbolic_ID<::__sycl_detail::__spec_id_shim_0()>() {
+      return "unique id for ::(unnamed)::same_name";
+    }
+  }
+}
+namespace {
+  namespace {
+    namespace __sycl_detail {
+      static constexpr decltype(same_name) &spec_id_shim_1() {
+        // address of ::(unnamed)::(unnamed)::same_name;
+        return same_name;
+      }
+    }
+  }
+
+  namespace __sycl_detail {
+    // Sometimes we need a 'shim', which points to another 'shim' in order to
+    // "extract" a variable from an anonymous namespace unambiguosly
+    static constexpr decltype(__sycl_detail::__spec_id_shim_1()) &__spec_id_shim_2() {
+      // still address of ::(unnamed)::(unnamed)::same_name;
+      return __sycl_detail::__spec_id_shim_1();
+    }
+  }
+}
+namespace sycl {
+  namespace detail {
+    template<>
+    inline const char *get_spec_constant_symbolic_ID<::__sycl_detail::__spec_id_shim_2()>() {
+      return "unique id for ::(unnamed)::(unnamed)::same_name";
+    }
+  }
+}
+
+#include <CL/sycl/detail/spec_const_integration.hpp>
+```
+
+### DPC++ runtime
+
+For each device binary compiler generates a map
+\<Symbolic ID\> =\> \<list of spec constant descriptors\> ("ID map"). DPC++
+runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
```

Comment:
Is the key the same thing in these two maps?  The ID map says the key is "\<Symbolic ID\>" while the value maps says the key is "\<Spec const symbolic ID\>".  If you intend for the key to be the same in both maps, use the description for both keys,  (For example, use "\<Symbolic ID\>" in both cases.)

Reply:
Yeah, those should refer to the same key, fixed in 7f64a62edc492777694860e91e4af9ae6cc26a6c